src/bundle/EventListener/UserCheckRequestListener.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Bundle\Rest\EventListener;
+
+use Ibexa\Bundle\Rest\Exception\UnexpectedUserException;
+use Ibexa\Contracts\Core\Repository\PermissionResolver;
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Security\Core\Security;
+
+final class UserCheckRequestListener implements EventSubscriberInterface, LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    private PermissionResolver $permissionResolver;
+
+    private Security $security;
+
+    public function __construct(PermissionResolver $permissionResolver, Security $security)
+    {
+        $this->permissionResolver = $permissionResolver;
+        $this->security = $security;
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            KernelEvents::REQUEST => [
+                ['checkUser'],
+            ],
+        ];
+    }
+
+    public function checkUser(RequestEvent $event): void
+    {
+        if (!$event->isMainRequest()) {
+            return;
+        }
+
+        if (!$event->getRequest()->attributes->get('is_rest_request')) {
+            return;
+        }
+
+        $request = $event->getRequest();
+        $expectedUserIdentifier = $request->headers->get('X-Expected-User');
+        if (empty($expectedUserIdentifier)) {
+            return;
+        }
+
+        $user = $this->security->getUser();
+        if ($user === null) {
+            throw new UnexpectedUserException('Expectation failed. User changed.', 401);
+        }
+
+        if (method_exists($user, 'getUserIdentifier')) {
```

Comment:
Why do we need `method_exists` check here?

Reply:
Because Symfony's `UserInterface` is changing between 5.x and 6.x. This ensures that both versions will work without changes.

I can drop it if we want this code to only support 5.x and we would just need to fix it going into 6.x.

Reply:
@adamwojs dropped the method check with expectation that we will swap to the other method for 6.x Symfony.